Source/DafnyCore/Resolver/PreType/PreType.cs

```diff
@@ -0,0 +1,326 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using Bpl = Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+  public abstract class PreType {
+    public PreType Normalize() {
+      var t = this;
+      while (t is PreTypeProxy proxy && proxy.PT != null) {
+        t = proxy.PT;
+      }
+      return t;
+    }
+
+    public static bool Same(PreType a, PreType b) {
+      a = a.Normalize();
+      b = b.Normalize();
+      if (a is DPreType ap && b is DPreType bp && ap.Decl == bp.Decl) {
+        Contract.Assert(ap.Arguments.Count == bp.Arguments.Count);
+        for (var i = 0; i < ap.Arguments.Count; i++) {
+          if (!Same(ap.Arguments[i], bp.Arguments[i])) {
+            return false;
+          }
+        }
+        return true;
+      }
+      return a == b;
+    }
+
+    public PreType UrAncestor(PreTypeResolver ptResolver) {
```

Comment:
What does `Ur` mean?

Reply:
From Oxford Dictionary:

prefix: ur-
primitive, original, or earliest.
Example: "urtext"

The dictionary says the prefix has a German origin. I think the Dutch prefix may be "ur-" or "oer-".

In short, I was trying to say "earliest ancestor" or "most primitive base type".

Reply:
As a native German speaker, I can confirm.